cdap-metrics-sidecar/src/main/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollector.java

```diff
@@ -0,0 +1,158 @@
+ * This service requires {@code SERVICE_NAME}  env variable to be set which provides
+ * the component tag for setting metrics context and the JMX server port number to be present in {@code cdap-site.xml}.
+ */
+public class JMXMetricsCollector extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(JMXMetricsCollector.class);
+  private static final long MEGA_BYTE = 1024 * 1024, MAX_PORT = (1 << 16) - 1;
+  private static final String SERVICE_URL_FORMAT = "service:jmx:rmi:///jndi/rmi://%s:%s/jmxrmi";
+  private final String componentName;
+  private final CConfiguration cConf;
+  private final MetricsCollectionService metricsCollectionService;
+  private final JMXServiceURL serviceUrl;
+  private final Environment env;
+  private ScheduledExecutorService executor;
+
+  @Inject
+  public JMXMetricsCollector(CConfiguration cConf,
+                             MetricsCollectionService metricsCollectionService,
+                             Environment env) throws MalformedURLException {
+    this.cConf = cConf;
+    int serverPort = cConf.getInt(Constants.JMXMetricsCollector.SERVER_PORT);
+    if (serverPort < 0 || serverPort > MAX_PORT) {
+      throw new IllegalArgumentException(String.format(
+        "%s variable (%d) is not a valid port number.",
+        Constants.JMXMetricsCollector.SERVER_PORT, serverPort));
+    }
+    String serverUrl = String.format(SERVICE_URL_FORMAT, "localhost", serverPort);
+    this.componentName = env.getVariable(Constants.JMXMetricsCollector.COMPONENT_NAME_ENV_VAR);
+    if (componentName == null) {
+      throw new IllegalArgumentException(
+        "Not collecting resource usage metrics from JMX as SERVICE_NAME env variable is not set.");
+    }
+    this.metricsCollectionService = metricsCollectionService;
+    this.env = env;
+    this.serviceUrl = new JMXServiceURL(serverUrl);
+  }
+
+  @Override
+  protected void startUp() {
+    LOG.info(String.format("Starting JMXMetricsCollector."));
+  }
+
+  @Override
+  protected void shutDown() {
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+    LOG.info(String.format("Shutting down JMXMetricsCollector has completed."));
+  }
+
+  @Override
+  protected void runOneIteration() {
+    Map<String, String> metricsContext = ImmutableMap.of(
+      Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+      Constants.Metrics.Tag.COMPONENT, componentName);
+    MetricsCollector metrics = this.metricsCollectionService.getContext(metricsContext);
+
+    try (JMXConnector jmxConnector = JMXConnectorFactory.connect(serviceUrl, null)) {
+      MBeanServerConnection mBeanConn = jmxConnector.getMBeanServerConnection();
+      getAndPublishMemoryMetrics(mBeanConn, metrics);
+      getAndPublishCPUMetrics(mBeanConn, metrics);
+      getAndPublishThreadMetrics(mBeanConn, metrics);
+    } catch (Exception e) {
+      LOG.error(String.format("Error occurred while connecting to JMX server."), e);
```

Comment:
Don't use `String.format` in log. Use `{}` placeholder instead. Also, this one seems doesn't even have any parameter, why need to use `String.format`? Please fix other usages as well.

Reply:
Changes made for other Logs as well.